File: user-services/internal/api/services/outbox_service.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"
	"user-services/internal/api/repositories"
	"user-services/internal/models"

	amqp "github.com/rabbitmq/amqp091-go"

	"github.com/google/uuid"
)

type OutboxService interface {
	PublishUserEvent(ctx context.Context, aggregateID uuid.UUID, eventType string, payload map[string]any) error
	ProcessUnpublishedEvents(ctx context.Context, limit int) error
	CleanupPublishedEvents(ctx context.Context) error
}

type outboxService struct {
	outboxRepo repositories.OutboxRepository
	channel    *amqp.Channel
	exchange   string
}

func NewOutboxService(outboxRepo repositories.OutboxRepository, channel *amqp.Channel, exchange string) OutboxService {
	return &outboxService{
		outboxRepo: outboxRepo,
		channel:    channel,
		exchange:   exchange,
	}
}

func (s *outboxService) PublishUserEvent(ctx context.Context, aggregateID uuid.UUID, eventType string, payload map[string]any) error {
	event := &models.Outbox{
		AggregateID: aggregateID,
		Topic:       "user.events",
		Type:        eventType,
		Payload:     payload,
		CreatedAt:   time.Now(),
	}
	return s.outboxRepo.Create(ctx, event)
}

func (s *outboxService) ProcessUnpublishedEvents(ctx context.Context, limit int) error {
	// 1. Fetch unpublished events
	events, err := s.outboxRepo.GetUnpublished(ctx, limit)
	if err != nil {
		return fmt.Errorf("failed to get unpublished events: %w", err)
	}

	if len(events) == 0 {
		return nil
	}

	log.Printf("Processing %d unpublished events", len(events))

	// 2. Publish each event to RabbitMQ
	for _, event := range events {
		if err := s.publishToRabbitMQ(ctx, event); err != nil {
			log.Printf("Failed to publish event %d: %v", event.ID, err)
			// Continue with other events instead of failing all
			continue
		}

		// 3. Mark as published
		if err := s.outboxRepo.MarkAsPublished(ctx, event.ID); err != nil {
			log.Printf("Failed to mark event %d as published: %v", event.ID, err)
			// Event was published but not marked - it will be retried
			// This is acceptable for at-least-once delivery
		}
	}

	return nil
}

func (s *outboxService) publishToRabbitMQ(ctx context.Context, event models.Outbox) error {
	// Create routing key from topic
	routingKey := event.Topic // e.g., "user.created", "user.events"

	// Marshal payload to JSON
	body, err := json.Marshal(event.Payload)
	if err != nil {
		return fmt.Errorf("failed to marshal event payload: %w", err)
	}

	// Publish to RabbitMQ
	err = s.channel.PublishWithContext(
		ctx,
		s.exchange, // exchange
		routingKey, // routing key
		false,      // mandatory
		false,      // immediate
		amqp.Publishing{
			ContentType:  "application/json",
			Body:         body,
			DeliveryMode: amqp.Persistent,
			Timestamp:    time.Now(),
			Type:         event.Type,
			Headers: amqp.Table{
				"aggregate_id": event.AggregateID.String(),
				"event_type":   event.Type,
			},
		},
	)

	if err != nil {
		return fmt.Errorf("failed to publish to RabbitMQ: %w", err)
	}

	log.Printf("Published event %d (type=%s) to exchange=%s with routing_key=%s",
		event.ID, event.Type, s.exchange, routingKey)

	return nil
}

func (s *outboxService) CleanupPublishedEvents(ctx context.Context) error {
	// Delete published events older than 7 days
	// Get the last event ID before cutoff
	events, err := s.outboxRepo.GetUnpublished(ctx, 1)
	if err != nil {
		return err
	}

	if len(events) > 0 {
		// Delete all published events with ID less than this
		return s.outboxRepo.DeletePublished(ctx, events[0].ID)
	}

	return nil
}

File: lesson-services/app/services/outbox_service.py
from sqlalchemy.orm import Session
from typing import List, Optional, Dict, Any
from uuid import UUID
from datetime import datetime, timedelta

# from app.models.progress_models import Outbox
# from app.schemas.outbox_schema import OutboxCreate

class OutboxService:
    """
    Outbox Pattern Service for reliable event publishing
    
    This service implements the Transactional Outbox pattern to ensure
    events are reliably published to message queue even if publishing fails.
    
    Flow:
    1. When domain event occurs, write to outbox table in same transaction
    2. Background worker polls outbox for unpublished messages
    3. Worker publishes to RabbitMQ and marks as published
    4. Old published messages are periodically cleaned up
    """
    
    def __init__(self, db: Session):
        self.db = db
    
    # get_pending_messages(limit: int = 100) -> List[Outbox]
    # Logic: Get unpublished messages for processing
    # - Query outbox WHERE published_at IS NULL
    # - Order by created_at ASC (FIFO - first in, first out)
    # - Limit to batch size
    # - Return list of pending messages
    # - Called by background worker polling
    
    # get_message(outbox_id: int) -> Optional[Outbox]
    # Logic: Get specific outbox message by ID
    # - Query outbox by id
    # - Return message or None
    
    # create_message(aggregate_id: UUID, topic: str, event_type: str, payload: Dict[str, Any]) -> Outbox
    # Logic: Create outbox message in same transaction as domain event
    # - Create outbox record with:
    #   - aggregate_id: ID of aggregate (user_id, lesson_id, etc.)
    #   - topic: message queue topic/exchange ('user.progress', 'lesson.completed')
    #   - type: event type ('LessonCompleted', 'QuizSubmitted', 'StreakUpdated')
    #   - payload: JSONB event data
    #   - created_at: current timestamp
    #   - published_at: NULL (unpublished)
    # - Insert into database
    # - DO NOT commit here - let caller commit with domain transaction
    # - Return created outbox record
    
    # mark_as_published(outbox_id: int) -> Optional[Outbox]
    # Logic: Mark message as successfully published
    # - Find outbox record by id
    # - Update published_at = current timestamp
    # - Commit transaction
    # - Return updated record
    # - Called by worker after successful publish to queue
    
    # mark_batch_as_published(outbox_ids: List[int]) -> int
    # Logic: Mark multiple messages as published (bulk operation)
    # - Update outbox SET published_at = now() WHERE id IN outbox_ids
    # - Commit transaction
    # - Return count of updated records
    # - More efficient for batch processing
    
    # cleanup_old_published(days_old: int = 7) -> int
    # Logic: Delete old published messages to prevent table bloat
    # - Calculate cutoff date (now - days_old)
    # - Delete outbox WHERE published_at < cutoff AND published_at IS NOT NULL
    # - Never delete unpublished messages
    # - Commit transaction
    # - Return count of deleted records
    # - Run periodically (e.g., daily cron job)
    
    # get_outbox_stats() -> Dict[str, Any]
    # Logic: Get statistics about outbox state
    # - Count pending messages (published_at IS NULL)
    # - Count published in last 24h
    # - Count old unpublished (created_at > 1 hour ago, still NULL)
    # - Calculate average publish time (published_at - created_at)
    # - Group pending by topic
    # - Return comprehensive statistics dict
    
    # get_failed_messages(age_hours: int = 1) -> List[Outbox]
    # Logic: Identify messages that failed to publish
    # - Query outbox WHERE published_at IS NULL AND created_at < (now - age_hours)
    # - These are stuck/failed messages
    # - Order by created_at ASC
    # - Return list for retry or investigation
    
    # retry_failed_message(outbox_id: int) -> Optional[Outbox]
    # Logic: Retry publishing a failed message
    # - Get outbox message
    # - Attempt to publish to message queue
    # - If successful, mark as published
    # - If failed, keep as unpublished for next retry
    # - Return updated message or None
    
    # get_messages_by_aggregate(aggregate_id: UUID) -> List[Outbox]
    # Logic: Get all outbox messages for specific aggregate
    # - Query outbox by aggregate_id
    # - Order by created_at DESC
    # - Return list of messages for audit/debugging
    
    # publish_message_to_queue(message: Outbox) -> bool
    # Logic: Actually publish message to RabbitMQ
    # - Extract topic, type, payload from outbox record
    # - Format as message queue event
    # - Publish to appropriate exchange/topic
    # - Return True if successful, False otherwise
    # - This is the actual integration with message queue
    # - Called by background worker
    
    # process_outbox_batch(batch_size: int = 100) -> int
    # Logic: Process a batch of outbox messages (worker main loop)
    # - Get pending messages (limit = batch_size)
    # - For each message:
    #   - Try to publish to queue
    #   - If successful, mark as published
    #   - If failed, log error and skip (will retry next time)
    # - Return count of successfully published messages
    # - This is the main worker function

